pkg/apis/walruscore/v1/resource_components.go

```diff
@@ -0,0 +1,130 @@
+package v1
+
+import (
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// ResourceComponent is the schema for the resource component API.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:crd-gen:resource:scope="Namespaced",subResources=["status"]
+
+type ResourceComponent struct {
+	meta.TypeMeta   `json:",inline"`
+	meta.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ResourceComponentsSpec   `json:"spec,omitempty"`
+	Status ResourceComponentsStatus `json:"status,omitempty"`
+}
+
+var _ runtime.Object = (*ResourceComponent)(nil)
+
+// ResourceComponentsSpec defines the desired state of ResourceComponents.
+type ResourceComponentsSpec struct{}
+
+// ResourceComponentsStatus defines the observed state of ResourceComponents.
+type ResourceComponentsStatus struct {
+	// Classes stores the component classes of the resource components.
+	Classes []ComponentClass `json:"classes"`
+}
+
+// ResourceComponentList holds the list of ResourceComponents.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ResourceComponentList struct {
+	meta.TypeMeta `json:",inline"`
+	meta.ListMeta `json:"metadata,omitempty"`
+
+	Items []ResourceComponent `json:"items"`
+}
+
+// ComponentMeta is the metadata of a component.
+type ComponentMeta struct {
+	// Name is the name of the component.
+	Name string `json:"name"`
+
+	// Type is the type of the component, example: "aws_instance" for aws_instance.foo.
+	Type string `json:"type"`
+
+	// Mode is the mode of the component.
+	Mode string `json:"mode"`
+}
+
+// ComponentClass is class resource components. The class defines the group of components instances.
+type ComponentClass struct {
+	ComponentMeta `json:",inline"`
+
+	// ConnectorRef is a reference to a connector.
+	ConnectorRef *ConnectorRef `json:"connectorRef,omitempty"`
+
+	// Instances stores component instances of the class resource components.
+	Instances []ComponentInstance `json:"instances"`
+}
+
+// ComponentInstance is instance resource components,
+// which is the realization of the class resource components.
+type ComponentInstance struct {
+	ComponentMeta `json:",inline"`
+
+	// The index key for any resource components that have been created using
+	// "count" or "for_each" combine with the module and resource name. e.g. `helm_release.foo[0]`.
+	IndexKey string `json:"indexKey"`
```

Comment:
should be blank?

Reply:
If it is not generated from count/for_each, we will fill it with default value `helm_release.foo`, blank is not expected.